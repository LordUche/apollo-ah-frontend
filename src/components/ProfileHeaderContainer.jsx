import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { toast } from 'react-toastify';
import { createStructuredSelector } from 'reselect';
import * as selectors from '../selectors/profileSelector';
import ProfileHeader from '../views/common/ProfileHeader';
import CloudinaryWidget from './common/CloudinaryWidget';
import { updateUserProfile, updatingProfile } from '../actions/profileAction';

let widget;

/**
 * Container component for the ProfileHeader view
 * @export
 * @class ProfileHeaderContainer
 * @extends {Component}
 */
class ProfileHeaderContainer extends Component {
  /**
   * Update the profile image on the server
   * @param {string} image The image url returned by cloudinary
   * @memberof ProfileHeaderContainer
   */
   updateProfileImage = async (image) => {
     const { profile, updateProfile, token } = this.props;
     const payload = { ...profile, token, image };
     await updateProfile(payload);
     const { errorData } = this.props;
     if (errorData.length !== 0) {
       toast.error(errorData[0]);
     }
   }

   /**
   * Opens the cloudinary image widget
   * @memberof ProfileHeaderContainer
   */
   handleImageUpload = () => {
     const { showLoader } = this.props;
     showLoader(true);
     if (widget === undefined) {
       widget = new CloudinaryWidget(this.updateProfileImage, showLoader, true);
     }
     widget.open();
   }

   render() {
     const { activePage, profile, isLoading } = this.props;
     return (
       <ProfileHeader
      activePage={activePage}
      profile={profile}
      isLoading={isLoading}
      handleImageUpload={this.handleImageUpload}
      />
     );
   }
}
ProfileHeaderContainer.propTypes = {
  activePage: PropTypes.string.isRequired,
  isLoading: PropTypes.bool,
  updateProfile: PropTypes.func.isRequired,
  token: PropTypes.string.isRequired,
  profile: PropTypes.object.isRequired,
  errorData: PropTypes.array,
  showLoader: PropTypes.func.isRequired,
};

ProfileHeaderContainer.defaultProps = {
  isLoading: false,
  errorData: []
};

const mapStateToProps = createStructuredSelector(
  {
    isLoading: selectors.getUpdatingProfile,
    token: selectors.getToken,
    profile: selectors.getProfile,
    errorData: selectors.getErrorData
  }
);
const mapDispatchToProps = dispatch => ({
  updateProfile: payload => dispatch(updateUserProfile(payload)),
  showLoader: (status = false) => dispatch(updatingProfile(status))
});
export default connect(mapStateToProps, mapDispatchToProps)(ProfileHeaderContainer);
