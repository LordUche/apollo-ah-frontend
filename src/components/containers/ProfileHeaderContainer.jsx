import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ProfileHeader from '../views/common/ProfileHeader';
import { updateUserProfile, setIsLoading } from '../../actions/profileAction';
import { UPDATING_PROFILE_IMAGE } from '../../actions/actionTypes';
import CloudinaryWidget from '../common/CloudinaryWidget';

let widget;

/**
 * Container component for the ProfileHeader view
 * @export
 * @class ProfileHeaderContainer
 * @extends {Component}
 */
class ProfileHeaderContainer extends Component {
  constructor(props) {
    super(props);
    this.handleImageUpload = this.handleImageUpload.bind(this);
    this.updateProfileImage = this.updateProfileImage.bind(this);
  }

  /**
   * Update the profile image on the server
   * @param {string} imageUrl The image url returned by cloudinary
   * @memberof ProfileHeaderContainer
   */
  updateProfileImage(imageUrl) {
    const { profile, updateProfile } = this.props;
    profile.image = imageUrl;
    updateProfile(profile);
  }

  /**
   * Opens the cloudinary image widget
   * @memberof ProfileHeaderContainer
   */
  handleImageUpload() {
    const { showLoader } = this.props;
    showLoader(true);
    if (widget === undefined) {
      widget = new CloudinaryWidget(this.updateProfileImage, showLoader, true);
    }
    widget.open();
  }

  render() {
    const { activePage, profile, isLoading } = this.props;
    return (
      <ProfileHeader
      activePage={activePage}
      profile={profile}
      isLoading={isLoading}
      handleImageUpload={this.handleImageUpload}
      />
    );
  }
}
ProfileHeaderContainer.propTypes = {
  showLoader: PropTypes.func.isRequired,
  activePage: PropTypes.string.isRequired,
  isLoading: PropTypes.bool.isRequired,
  updateProfile: PropTypes.func.isRequired,
  profile: PropTypes.shape({
    firstname: PropTypes.string.isRequired,
    lastname: PropTypes.string.isRequired,
    username: PropTypes.string.isRequired,
    bio: PropTypes.string.isRequired,
  }).isRequired,
};

const mapStateToProps = state => ({
  profile: state.user,
  isLoading: state.loading.updatingProfileImage,
});
const mapDispatchToProps = dispatch => ({
  updateProfile: userData => dispatch(updateUserProfile(UPDATING_PROFILE_IMAGE, userData)),
  showLoader: (status = false) => dispatch(setIsLoading(UPDATING_PROFILE_IMAGE, status))
});
export default connect(mapStateToProps, mapDispatchToProps)(ProfileHeaderContainer);
