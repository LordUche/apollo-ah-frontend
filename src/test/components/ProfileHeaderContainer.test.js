import React from 'react';
import setup from '../setup';
import ProfileHeaderContainer from '../../components/ProfileHeaderContainer';

const container = <ProfileHeaderContainer activePage="PROFILE" />;
const mockFn = jest.fn();

window.cloudinary = {
  createUploadWidget: () => ({
    open: mockFn,
  }),
};

describe('<ProfileHeaderContainer>', () => {
  it('should render without crashing', () => {
    const wrapper = setup(container);
    expect(wrapper).toMatchSnapshot();
  });
  it('should open cloudinary widget', () => {
    const wrapper = setup(container);
    expect(wrapper.find('.change-image').simulate('click', mockFn));
    expect(window.cloudinary.createUploadWidget().open).toHaveBeenCalled();
  });
});
