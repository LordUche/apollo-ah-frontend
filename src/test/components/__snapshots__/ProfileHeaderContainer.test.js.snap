// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfileHeaderContainer> should render without crashing 1`] = `ReactWrapper {}`;
